### DISCUSSION TITLE
How can I emulate Liquid’s `where` on Nunjucks?

### DISCUSSION THREAD

**Discussion by ed-ponce** [11ty/eleventy#1776] [Q&A] [answered]
Using [Liquid’s `where`](https://shopify.github.io/liquid/filters/where) filter you can very easily subset an array’s content. For example, to list posts that are not a draft: `{{ collections.posts | where: "data.draft", false }}`.

I’m trying to use Nunjucks’ `reject` and `rejectattr` but I’ve had no luck. `{{ collections.posts | rejectattr("data.draft") }}` renders the whole array, with no effect.

Then I attempted to write a new filter, but my JavaScript is very very limited. This is what I have:

```
eleventyConfig.addNunjucksFilter("where", function(arr, foo, bar) {
  return arr.filter((entry) => http://entry.data.foo == bar );
});
```

Then, you can hard-code a custom, filtered collection within `.eleventy.js`, but I find it a bit impractical because I used to use `where` to create very different arrays on the fly. Also, I wouldn’t know how to write such collection.

**Comment by pdehaan** [comment, marked as answer]
I haven't tried using Nunjucks' [`rejectattr`](https://mozilla.github.io/nunjucks/templating.html#rejectattr-only-the-single-argument-form) filter yet, but this _might_ get you somewhat closer:

```js
// Useful for trying to compare nested properties...
const _get = require("lodash.get");

module.exports = (eleventyConfig) => {
  eleventyConfig.addNunjucksFilter("where", (arr, key, expected) => {
    if (expected === undefined) {
      return arr.filter(item => !!_get(item, key));
    }
    return arr.filter(item => _get(item, key) === expected);
  });
  // ...
}
```

```nunjucks
{% for post in collections.posts | where("data.draft", false) %}
  <p data-draft="{{ post.data.draft }}">{{ post.data.title }}</p>
{% else %}
  <p>No posts found</p>
{% endfor %}
```

Not sure that it will work exactly like [LiquidJS `where` implementation](https://github.com/harttle/liquidjs/blob/0acfb8de4dc90c1b55e25996d7e05fed3b17ed7e/src/builtin/filters/array.ts#L40-L45), or that my hacky filter doesn't have a few bugs... 
I had to add a posts.json directory data file to my ./src/posts/ folder to set the default value for `draft` to false, and then override it to `true` in the front matter for a couple templates.

```json
{
  "draft": false,
  "tags": ["posts"]
}
```

**Reply by ed-ponce** [reply]
That works for this specific case! The property has to be defined with front matter defaults, like you said. Thank you!

However, if I try to, for example, target an array like tags with something like `where("tags", "tutorial")` (as in “containing `tutorial` within the array”), it wouldn’t work.

**Reply by pdehaan** [reply]
No, but I don't think that'd work in LiquidJS's filter implementation either. I'd probably write a separate filter for that.

```nunjucks
{% for post in collections.posts | postIncludesTag("eel") %}
  <p data-draft="{{ post.data.tags }}">{{ post.data.title }}</p>
{% else %}
  <p>No eel based blog posts</p>
{% endfor %}
```

```js
eleventyConfig.addFilter("postIncludesTag", (arr, tag) => {
  return arr.filter(item => {
    // console.log({title: item.data.title, tags: item.data.tags, tag});
    return item.data.tags?.includes(tag);
  });
});
```

**Reply by pdehaan** [reply]
Oh, and I was trying to play around with it a bit more... I think I found a solution for not needing to explicitly set `draft: false` on all the posts using the directory data file. By explicitly checking for an `expected === null` (instead of `undefined`) as a default in the filter, I'm now able to pass `undefined` in my `where` filter to check for all blog posts where `post.data.draft` isn't explicitly defined.

```js
eleventyConfig.addNunjucksFilter("where", (arr, key, expected = null) => {
  if (expected === null) {
    return arr.filter(item => !!_get(item, key));
  }
  return arr.filter(item => _get(item, key) === expected);
});
```

```nunjucks
{% for post in collections.posts | where("data.draft", undefined) %}
  <p data-draft="{{ post.data.draft }}">{{ post.data.title }}</p>
{% else %}
  <p>No posts found</p>
{% endfor %}
```

```json
{
  "tags": ["posts"]
}
```

**Reply by ed-ponce** [reply]
> ```js
> eleventyConfig.addFilter("postIncludesTag", (arr, tag) => {
>   return arr.filter(item => {
>     // console.log({title: item.data.title, tags: item.data.tags, tag});
>     return item.data.tags?.includes(tag);
>   });
> });
> ```
Thank you so so much for your help!

> ```js
> eleventyConfig.addNunjucksFilter("where", (arr, key, expected = null) => {
>   if (expected === null) {
>     return arr.filter(item => !!_get(item, key));
>   }
>   return arr.filter(item => _get(item, key) === expected);
> });
> ```
Is there a way to expect both `undefined` and `null` values just in case?

**Reply by pdehaan** [reply]
> Is there a way to expect both `undefined` and `null` values just in case?

You could possibly do something like this, but then I think it changes the logic around the default behavior where LiquidJS was trying to match for truthy values if you didn't pass an expected value:

```js
eleventyConfig.addNunjucksFilter("where", (arr, key, expected = null) => {
  if ([null, undefined].includes(expected)) {
    return arr.filter(item => !!_get(item, key));
  }
  return arr.filter(item => _get(item, key) === expected);
});
```

But now I can't see how to modify that to do the truthy default check AND check for instances where `data.draft` is `undefined`.

**Reply by ed-ponce** [reply]
Oooh it’s like you can’t have the three of them. Undefined values should be treated as falsy, shouldn’t they?

I’ll stay with the first implementation for now.

**Reply by pdehaan** [reply]
https://developer.mozilla.org/en-US/docs/Glossary/Falsy and https://developer.mozilla.org/en-US/docs/Glossary/Truthy do a pretty good job of explaining the default truthy-vs-falsy values in JavaScript. Although, LiquidJS and Nunjucks might have their own rules/implementations...
For example, https://liquidjs.com/tutorials/truthy-and-falsy.html.
I don't think Nunjucks has similar docs yet, but I filed https://github.com/mozilla/nunjucks/issues/1268 a while ago and there is an interesting link to the v4 roadmap and a couple interesting truthy related changes.

**Comment by ed-ponce** [comment]
`{% for post in collections.posts if not post.data.draft %}` wouldn’t work either, for some reason.

**Reply by pdehaan** [reply]
Interesting. Per https://github.com/mozilla/nunjucks/issues/261, I don't think Nunjucks supports that loop filtering syntax.
Looks like it might be on the roadmap for v4 though: https://github.com/mozilla/nunjucks/issues/1059 (although that was like a 3 year old issue).

**Comment by edwardhorsford** [comment]
I've got a where filter in on one of my projects:

```
// Filter results for only those containing attribute and value
exports.where = (arr, key, compare) => {
  compare = [].concat(compare) // force to arr
  let filtered = arr.filter(item => {
    return compare.includes(item[key])
  })
  return filtered
}

// Remove items with a specified attribute and value
exports.removeWhere = (arr, key, compare) => {
  compare = [].concat(compare) // force to arr
  let filtered = arr.filter(item => {
    return !compare.includes(item[key])
  })
  return filtered
}
```

Both of these also accept an array for the compare value. So you could do `{{ array | removeWhere('number', ['one', 'two']) }}`

**Reply by ed-ponce** [reply]
I’d love to try these. How can I implement them? I’m not familiar with this syntax.

**Comment by edwardhorsford** [comment]
Ah sorry - I store my filters in a separate file to my eleventy stuff.

Wrapped in eleventy syntax:
```
eleventyConfig.addNunjucksFilter("where", (arr, key, compare) => {
  compare = [].concat(compare) // force to arr
  let filtered = arr.filter(item => {
    return compare.includes(item[key])
  })
  return filtered
})

eleventyConfig.addNunjucksFilter("removeWhere", (arr, key, compare) => {
  compare = [].concat(compare) // force to arr
  let filtered = arr.filter(item => {
    return !compare.includes(item[key])
  })
  return filtered
})

```

**Reply by ed-ponce** [reply]
I tried this snippet with no luck. For example: I used `{% for entry in collections.all | removeWhere("data.permalink", false) %}` to filter pages with falsy `permalink` attributes out of the sitemap.

I couldn’t find a use where either filter had an effect.

What could be causing this?

**Reply by pdehaan** [reply]
I think you might need to use something like [lodash.get](http://npm.im/lodash.get) in order to pluck nested properties like "data.permalink".

This seemed to work for me locally:

```diff
+  const _get = require("lodash.get");
  // ...
  eleventyConfig.addNunjucksFilter("removeWhere", (arr, key, compare) => {
    compare = [].concat(compare); // force to arr
    let filtered = arr.filter(item => {
+      return !compare.includes(_get(item, key));
-      return !compare.includes(item[key])
    })
    return filtered;
  });
```

```njk
---
title: removeWhere Test
---

{% for entry in collections.all | removeWhere("data.permalink", false) %}
  title={{ entry.data.title }}; url={{ entry.url | url }}<br/>
{% else %}
  No entries??
{% endfor %}
```

That should exclude any collection items where the `permalink` front matter value is explicitly `false`. If you're trying to do a more complicated falsy `permalink` check, you might need some more complex logic.

**Reply by ed-ponce** [reply]
Now THAT worked. This is the closest I got to emulate Liquid’s `where` as I remember it. Plus, added functionality with `removeWhere`.

Thank you 🙏
